streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -336,22 +371,34 @@ public V delete(K key) {
 
     @Override
     public KeyValueIterator<K, V> range(K from, K to) {
-        // we need to flush the cache if necessary before returning the iterator
-        if (cache != null)
-            flushCache();
+        lock.readLock().lock();
+        try {
+            validateStoreOpen();
+            // we need to flush the cache if necessary before returning the iterator
+            if (cache != null)
+                flushCache();
 
-        return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
+            return new RocksDBRangeIterator<>(db.newIterator(), serdes, from, to);
```

Comment:
I did start down that path, but I opted against it as I was concerned about callers that grab an iterator and take a long time to iterate or never close it - in this case the `RocksDBStore` wouldn't be able to be closed until the iterator completes, which means rebalances have the potential to become blocked. 
So, i figured that if the store gets closed a call to next will throw an exception. IMO this is perfectly valid. Thoughts? @guozhangwang